tox.ini

```diff
@@ -29,6 +29,18 @@ commands =
     isort {[vars]all_path}
     black {[vars]all_path}
 
+[testenv:src-docs]
+allowlist_externals=sh
+setenv =
+    PYTHONPATH=src
+description = Generate documentation for src
+deps =
+    lazydocs
+    -r{toxinidir}/requirements.txt
+commands =
+    ; can't run lazydocs directly due to needing to run it on src/* which produces an invocation error in tox
```

Comment:
For my understanding, why should it run on "src/*" and not `src/` ?

Reply:
This is a bit messy because `lazydocs` expects a module whereas the charm isn't really a module (e.g., it doesn't have an `__init__.py`). To make `lazydocs` work, it is necessary to run from the root directory and add `src` to the `PYTHONPATH`. This results in pydantic seeing the `timerange.py` twice which results in an error around pydantic not allowing re-using root validators even though the code, as running in production, doesn't do this. One option is to add an argument to allow re-use, though it isn't required. This way, each module is treated independently rather than collectively, which is also more like what the charm does in production

Reply:
I spent a few hours trying to make this work, this is the best I could figure out